chat-bots/src/Control/Monad/ListT.hs

```diff
@@ -41,44 +45,76 @@ newtype ListT m a = ListT
   }
 
 instance Functor m => Functor (ListT m) where
-  fmap :: Functor m => (a -> b) -> ListT m a -> ListT m b
+  fmap :: (a -> b) -> ListT m a -> ListT m b
   fmap f (ListT ma) = ListT $ fmap (bimap f (fmap f)) $ ma
 
 instance Monad m => Applicative (ListT m) where
-  pure :: Monad m => a -> ListT m a
-  pure = ListT . return . (`ConsF` emptyListT)
+  pure :: a -> ListT m a
+  pure a = a <$ Functor.introduce @_ @() ()
 
-  (<*>) :: Monad m => ListT m (a -> b) -> ListT m a -> ListT m b
-  (<*>) = ap
+  liftA2 :: (a -> b -> c) -> ListT m a -> ListT m b -> ListT m c
+  liftA2 f xs ys = uncurry f <$> Functor.combine (xs, ys)
 
-instance Monad m => Alternative (ListT m) where
-  empty :: Monad m => ListT m a
-  empty = emptyListT
+instance Monad m => Functor.Semigroupal (->) (,) (,) (ListT m) where
+  combine :: (ListT m x, ListT m x') -> ListT m (x, x')
+  combine = uncurry (liftM2 (,))
+
+instance Monad m => Functor.Unital (->) () () (ListT m) where
+  introduce :: () -> ListT m ()
+  introduce () = ListT $ return (() `ConsF` emptyListT)
```

Comment:
We should probably provide newtypes to derive `Monoidal ...` from `Applicative`/`Alternative`/etc. and vice versa. It would be nice for it to be immediately obvious at a scan that we're just trivially porting over an instance of an equivalent typeclass.

Reply:
The `From*` newtypes exist but aren't exported. I can export them. What about leveraging generics for this? Would that be possible?